src/dagon/core/event.d

```diff
@@ -86,7 +88,19 @@ class EventManager
     bool running = true;
 
     bool[512] keyPressed = false;
+    bool[512] keyUp = false;
+    bool[512] keyDown = false;
+
     bool[255] mouseButtonPressed = false;
+    bool[255] mouseButtonUp = false;
+    bool[255] mouseButtonDown = false;
+
+    bool[255] controllerButtonPressed = false;
+    bool[255] controllerButtonUp = false;
+    bool[255] controllerButtonDown = false;
```

Comment:
Optimization tip: you can use `std::bitset` and it will save you some memory. Not sure, but maybe it even increase performance a little bit.

Reply:
You probably mean `std.bitmanip`? I think this is possible, thanks for a tip.

Reply:
Do we really care? It's using between 3KB to 24KB of memory(depends on bool size), and modern computers have GBs of memory. Performance wise I believe it would be slower because of bit masking operations involved in accessing specific bits but to be sure we would need to make some benchmarks.

Reply:
> You probably mean std.bitmanip? I think this is possible, thanks for a tip.

Oh... I thought this is C++... Oops